`versionCatalogUpdate` task failing on `Collection contains no element matching the predicate`

It's unclear what collection is even being referenced... also, the update to libs.versions.toml happens nevertheless!

```
> Task :versionCatalogUpdate FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':versionCatalogUpdate'.
> Collection contains no element matching the predicate.
```

Not much can be inferred from `--info` logs either.

Full repro: the current state of https://github.com/VirtusLab/git-machete-intellij-plugin/pull/1009, i.e. https://github.com/VirtusLab/git-machete-intellij-plugin/tree/bffe9f95ef0b32e674c209ea58b07e0243ebdc80

Thanks for the repro, that really helps a lot! Running `./gradlew versionCatalogUpdate --stacktrace` reveals where that exception is thrown:

```
Caused by: java.util.NoSuchElementException: Collection contains no element matching the predicate.
        at nl.littlerobots.vcu.plugin.VersionCatalogUpdateTask.checkForUpdatedPinnedPlugins(VersionCatalogUpdateTask.kt:589)
        at nl.littlerobots.vcu.plugin.VersionCatalogUpdateTask.updateCatalog(VersionCatalogUpdateTask.kt:154)
```

That's indeed after the catalog is updated for outputting information on pinned plugins. I'll need to dig in a bit deeper to see what condition is causing this still.

--- a/plugin/src/main/kotlin/nl/littlerobots/vcu/plugin/VersionCatalogUpdateTask.kt
+++ b/plugin/src/main/kotlin/nl/littlerobots/vcu/plugin/VersionCatalogUpdateTask.kt
@@ -172,9 +172,12 @@ abstract class VersionCatalogUpdateTask @Inject constructor() : DefaultTask() {
             resolvedVersions.libraries.values.any {
                 it.group == pin.group
             }
-        }.map { lib ->
-            lib to catalogWithResolvedPlugins.libraries.entries.first {
+        }.mapNotNull { lib ->
+            // can be null for kept, but unused libraries
+            catalogWithResolvedPlugins.libraries.entries.firstOrNull {
                 it.value.group == lib.group
+            }?.let {
+                lib to it
             }
         }.filter {
             it.first.version != it.second.value.version && it.first.version is VersionDefinition.Simple
@@ -217,9 +220,12 @@ abstract class VersionCatalogUpdateTask @Inject constructor() : DefaultTask() {
             resolvedVersions.plugins.values.any {
                 it.id == pin.id
             }
-        }.map { plugin ->
-            plugin to catalogWithResolvedPlugins.plugins.entries.first {
+        }.mapNotNull { plugin ->
+            // can be null for kept, but unused plugins
+            catalogWithResolvedPlugins.plugins.entries.firstOrNull() {
                 it.value.id == plugin.id
+            }?.let {
+                plugin to it
             }
         }.filter {
             it.first.version != it.second.value.version && it.first.version is VersionDefinition.Simple

--- a/plugin/src/test/kotlin/nl/littlerobots/vcu/plugin/VersionCatalogUpdatePluginTest.kt
+++ b/plugin/src/test/kotlin/nl/littlerobots/vcu/plugin/VersionCatalogUpdatePluginTest.kt
@@ -1280,4 +1280,47 @@ class VersionCatalogUpdatePluginTest {
             libs
         )
     }
+
+    @Test
+    // repro for https://github.com/littlerobots/version-catalog-update-plugin/issues/61
+    fun `unused pins should be ignored when generating pin warnings`() {
+        val reportJson = tempDir.newFile()
+
+        buildFile.writeText(
+            """
+            plugins {
+                id "nl.littlerobots.version-catalog-update"
+            }
+
+            tasks.named("versionCatalogUpdate").configure {
+                it.reportJson = file("${reportJson.name}")
+            }
+            """.trimIndent()
+        )
+
+        reportJson.writeText("{}")
+
+        val toml = """
+            [libraries]
+            # @pin
+            # @keep
+            aaa = "some:library:2.0"
+
+            [plugins]
+            # @pin
+            # @keep
+            aaa = "another.id:1.0.0"
+
+        """.trimIndent()
+
+        File(tempDir.root, "gradle").mkdir()
+        File(tempDir.root, "gradle/libs.versions.toml").writeText(toml)
+
+        GradleRunner.create()
+            .withProjectDir(tempDir.root)
+            .withArguments("versionCatalogUpdate")
+            .withDebug(true)
+            .withPluginClasspath()
+            .build()
+    }
 }

